### ISSUE TITLE
Missing a step in 22.2.5.13 RegExp.prototype [ @@split ]

### ISSUE THREAD

**Issue by vqun** [tc39/ecma262#2740]
As the latest spec describes in 22.2.5.13 RegExp.prototype [ @@split ]:

> // Dismiss 1-16 steps
> 
>17. Let p be 0.
>18. Let q be p.
>19. Repeat, while q < size,
>    a. Perform ? Set(splitter, "lastIndex", 𝔽(q), true).
>    b. Let z be ? RegExpExec(splitter, S).
>    c. If z is null, set q to AdvanceStringIndex(S, q, unicodeMatching).
>    d. Else,
>        i. Let e be ℝ(? ToLength(? Get(splitter, "lastIndex"))).
>        ii. Set e to min(e, size).
>        iii. If e = p, set q to AdvanceStringIndex(S, q, unicodeMatching).
>        iv. Else,
>          1. Let T be the substring of S from p to q.
>          2. Perform ! CreateDataPropertyOrThrow(A, ! ToString(𝔽(lengthA)), T).
>          3. Set lengthA to lengthA + 1.
>          4. If lengthA = lim, return A.
>          5. Set p to e.
>          6. Let numberOfCaptures be ? LengthOfArrayLike(z).
>          7. Set numberOfCaptures to max(numberOfCaptures - 1, 0).
>          8. Let i be 1.
>          9. Repeat, while i ≤ numberOfCaptures,
>              a. Let nextCapture be ? Get(z, ! ToString(𝔽(i))).
>              b. Perform ! CreateDataPropertyOrThrow(A, ! ToString(𝔽(lengthA)), nextCapture).
>              c. Set i to i + 1.
>              d. Set lengthA to lengthA + 1.
>              e. If lengthA = lim, return A.
>          10. Set q to p.
>20. Let T be the substring of S from p to size.
>21. Perform ! CreateDataPropertyOrThrow(A, ! ToString(𝔽(lengthA)), T).
>22. Return A.

For the step 19.d.iv.1:
>          1. Let T be the substring of S from p to q.

p and q are always same. But in fact in this step, it should push the unmatched string(substring from lastIndex to current matched index) to the result A. It should be such as:

> 0. Set q to ToLength(? Get(z, "index")) - 1
> Let T be the substring of S from p to q

I am confused, so please kindly help to disabuse if my saying is wrong. Thanks a lot!

**Comment by ljharb** [comment]
Reading https://tc39.es/ecma262/#sec-regexp.prototype-@@split, q is set on step 19.c when `z` is null. why would it be the same as p on 19.d.iv.1 in every case?

**Comment by vqun** [comment]
@ljharb step 19.c is performed only when z is null, does't that means `RegExpExec(splitter, S)` is 'failed' and S is not matched with spliter?

I am confused. For example:

```javascript
const str = 'ab2022cd'
const reg = /\d+/

const s = str.split(reg)
````

in the first loop, performing 19.b (`Let z be ? RegExpExec(splitter, S).`) will result `z is not null`, that means `19.c` is false, and will perform `19.d Else`, in such case, `p = q`, and `19.d.iv.1` will result an empty `T` which will be 'pushed' to the return `A`.

Please help disabusing me, thanks a lot!

**Comment by ljharb** [comment]
Right - which means that until `q` is `>= size`, or until it matches, it will advance the string index for `q` (making it not equal `p` if there's at least one failed match).

**Comment by vqun** [comment]
> @ljharb step 19.c is performed only when z is null, does't that means `RegExpExec(splitter, S)` is 'failed' and S is not matched with spliter?
> 
> I am confused. For example:
> 
> ```js
> const str = 'ab2022cd'
> const reg = /\d+/
> 
> const s = str.split(reg)
> ```
> 
> in the first loop, performing 19.b (`Let z be ? RegExpExec(splitter, S).`) will result `z is not null`, that means `19.c` is false, and will perform `19.d Else`, in such case, `p = q`, and `19.d.iv.1` will result an empty `T` which will be 'pushed' to the return `A`.
> 
> Please help disabusing me, thanks a lot!

as what I mentioned here, "**and `19.d.iv.1` will result an empty `T` which will be 'pushed' to the return `A`.**", but for the implementation of the browsers, A does not contain an empty string as the first element.(means `const s = str.split(reg)`, s[0] is not '').

Do I miss some key cases?

**Comment by Josh-Cena** [comment]
Just want to mention that I captured the same issue while reading this section. Just thinking about the trivial case of the first iteration since I'm still unclear what the semantics of `p`, `q` and `e` are: if the match result is not `null`, and `e != p` (i.e. the match is not at the very beginning), `p` and `q` would both be 0 when it reaches 19.d.iv.1. After that, `p` and `q` both seem to be set to `e`, which means... they are equal most of the time? Am I missing something or do I just don't understand how this works at all?

**Comment by vqun** [comment]
> Just want to mention that I captured the same issue while reading this section. Just thinking about the trivial case of the first iteration since I'm still unclear what the semantics of `p`, `q` and `e` are: if the match result is not `null`, and `e != p` (i.e. the match is not at the very beginning), `p` and `q` would both be 0 when it reaches 19.d.iv.1. After that, `p` and `q` both seem to be set to `e`, which means... they are equal most of the time? Am I missing something or do I just don't understand how this works at all?

Yes,  that's why I think the spec misses a step. Maybe we read it with incorrect way, and truly confuse me.
